pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,786 @@
+// it can be retrieved using only the UID.  If a Pod UID is given, only one set
+// of API credentials are allowed per pod.  All secret references are stored in
+// the default namespace but contain no sensitive data.
+func (plugin *storageosPlugin) storeSecretRef(podNamespace, secretName string, UID types.UID) (string, error) {
+	kubeClient := plugin.host.GetKubeClient()
+	if kubeClient == nil {
+		return "", fmt.Errorf("Cannot get kube client")
+	}
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: defaultNamespace,
+			Name:      secretRefName(UID),
+		},
+		Data: map[string][]byte{
+			"podNamespace": []byte(podNamespace),
+			"secretName":   []byte(secretName),
+		},
+		Type: storageosPluginName,
+	}
+	_, err := kubeClient.Core().Secrets(defaultNamespace).Create(secret)
+	if errors.IsAlreadyExists(err) {
+		err = nil
+	}
+	if err != nil {
+		return "", fmt.Errorf("Couldn't create secret %v/%v: %v", podNamespace, secretName, err)
+	}
+	return secretRefName(UID), err
+}
+
+// lookupSecretRef returns the location of the API credentials used by the pod
+// or PV.
+func (plugin *storageosPlugin) lookupSecretRefByID(UID types.UID) (string, string, error) {
+	return plugin.lookupSecretRef(secretRefName(UID))
+}
+
+func (plugin *storageosPlugin) lookupSecretRef(refName string) (string, string, error) {
+	kubeClient := plugin.host.GetKubeClient()
+	if kubeClient == nil {
+		return "", "", fmt.Errorf("Cannot get kube client")
+	}
+	secretMap, err := util.GetSecretForPV(defaultNamespace, refName, storageosPluginName, kubeClient)
+	if err != nil {
+		return "", "", err
+	}
+
+	var podNamespace, secretName string
+
+	for name, data := range secretMap {
+		if name == "podNamespace" {
+			podNamespace = string(data)
+		}
+		if name == "secretName" {
+			secretName = string(data)
+		}
+	}
+	return podNamespace, secretName, nil
+}
+
+// removeSecretRef deletes the secret that holds the location of the API creds
+// for the pod or PV.
+func (plugin *storageosPlugin) removeSecretRef(UID types.UID) error {
+	kubeClient := plugin.host.GetKubeClient()
+	if kubeClient == nil {
+		return fmt.Errorf("Cannot get kube client")
+	}
+	return kubeClient.Core().Secrets(defaultNamespace).Delete(secretRefName(UID), &metav1.DeleteOptions{})
+}
+
+// secretRefName returns a prefixed name based on the pod or PV UID so that it
+// will be unique within the default namespace.
+func secretRefName(UID types.UID) string {
+	return fmt.Sprintf("%s-%v", secretLocatorPrefix, UID)
+}
+
+func (plugin *storageosPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volume.VolumeOptions) (volume.Mounter, error) {
+	source, _, err := getVolumeSource(spec)
+	if err != nil {
+		glog.Warningf("getVolumeSource err: %v", err)
+		return nil, err
+	}
+
+	var secretNamespace, secretName string
+
+	// SecretName will be a locator ref for dynamically provisioned volumes
+	if strings.HasPrefix(source.SecretName, secretLocatorPrefix) {
```

Comment:
this doesn't seem like a reliable way to plumb this info, especially since it means the secret can be fetched from a namespace other than the pod namespace